Deriving BoundedArbitrary fails for enum with one variant

<!--
If this is a security issue, please report it following the
[security reporting procedure](https://github.com/model-checking/kani/security/policy).
-->

<!--
Thank you for filing a bug report! 🐛
Please provide a short summary of the issue, along with the information necessary to replicate.
-->

I tried this code:
<!--
If the reproducing code is small, please post it here.
Otherwise, please attach the relevant files to this issue.
Alternatively, post a link to the repository and branch that exposes the issue.
-->
```rust
#[derive(kani::BoundedArbitrary)]
enum Foo {
    A(#[bounded] String),
    //B(#[bounded] String),
}

#[kani::proof]
fn check_enum() {
    let _foo = kani::bounded_any::<Foo, 4>();
}
```

using the following command line invocation:

```
kani bounded_enum.rs
```

with Kani version: 37fb6345740

I expected to see this happen: Verification successful

<!--
If Kani crashed, please include a backtrace in the code block by setting
`RUST_BACKTRACE=1` in your environment. E.g. `RUST_BACKTRACE=1 kani ...`
-->
Instead, this happened:
```bash
 $ kani bounded_enum.rs 
Kani Rust Verifier 0.63.0 (standalone)
error[E0283]: type annotations needed
  --> bounded_enum.rs:1:10
   |
1  | #[derive(kani::BoundedArbitrary)]
   |          ^^^^^^^^^^^^^^^^^^^^^^ cannot infer type
   |
   = note: cannot satisfy `_: kani::Arbitrary`
   = help: the following types implement trait `kani::Arbitrary`:
             ()
             (A, B)
             (A, B, C)
             (A, B, C, D)
             (A, B, C, D, E)
             (A, B, C, D, E, F)
             (A, B, C, D, E, F, G)
             (A, B, C, D, E, F, G, H)
           and 50 others
note: required by a bound in `kani::any`
  --> /home/ubuntu/git/kani/library/kani/src/lib.rs:55:1
   |
55 | kani_core::kani_lib!(kani);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `any`
   = note: this error originates in the derive macro `kani::BoundedArbitrary` which comes from the expansion of the macro `kani_core::kani_lib` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0283`.
error: /home/ubuntu/git/kani/target/kani/bin/kani-compiler exited with status exit status: 101
```

If I uncomment the second variant, verification is successful.

--- a/library/kani_macros/src/derive_bounded.rs
+++ b/library/kani_macros/src/derive_bounded.rs
@@ -92,19 +92,40 @@ fn generate_type_constructor(type_name: TokenStream, fields: &syn::Fields) -> To
 /// Generates a `match` case to construct each variant of the given type. Uses a
 /// symbolic `usize` to decide which variant to construct.
 fn enum_constructor(ident: &syn::Ident, data_enum: &syn::DataEnum) -> TokenStream {
-    let variant_constructors = data_enum.variants.iter().map(|variant| {
+    if data_enum.variants.is_empty() {
+        let msg = format!(
+            "Cannot create symbolic enum `{ident}`. Enums with zero-variants cannot be instantiated"
+        );
+        quote! {
+            panic!(#msg)
+        }
+    } else if data_enum.variants.len() == 1 {
+        let variant = data_enum.variants.first().unwrap();
         let variant_name = &variant.ident;
-        generate_type_constructor(quote!(#ident::#variant_name), &variant.fields)
-    });
-    let n_variants = data_enum.variants.len();
-    let cases = variant_constructors.enumerate().map(|(idx, var_constr)| {
-        if idx < n_variants - 1 { quote!(#idx => #var_constr) } else { quote!(_ => #var_constr) }
-    });
+        let variant_constructor =
+            generate_type_constructor(quote!(#ident::#variant_name), &variant.fields);
+        quote! {
+            #variant_constructor
+        }
+    } else {
+        let variant_constructors = data_enum.variants.iter().map(|variant| {
+            let variant_name = &variant.ident;
+            generate_type_constructor(quote!(#ident::#variant_name), &variant.fields)
+        });
+        let n_variants = data_enum.variants.len();
+        let cases = variant_constructors.enumerate().map(|(idx, var_constr)| {
+            if idx < n_variants - 1 {
+                quote!(#idx => #var_constr)
+            } else {
+                quote!(_ => #var_constr)
+            }
+        });
 
-    let kani_path = kani_path();
-    quote! {
-        match #kani_path::any() {
-            #(#cases),* ,
+        let kani_path = kani_path();
+        quote! {
+            match #kani_path::any() {
+                #(#cases),* ,
+            }
         }
     }
 }

--- a/tests/expected/derive-bounded-arbitrary/enum_one_variant.expected
+++ b/tests/expected/derive-bounded-arbitrary/enum_one_variant.expected
@@ -1,0 +1,7 @@
+Checking harness check_enum...
+
+ ** 5 of 5 cover properties satisfied
+
+VERIFICATION:- SUCCESSFUL
+
+Complete - 1 successfully verified harnesses, 0 failures, 1 total.

--- a/tests/expected/derive-bounded-arbitrary/enum_one_variant.rs
+++ b/tests/expected/derive-bounded-arbitrary/enum_one_variant.rs
@@ -1,0 +1,23 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! Check that derive BoundedArbitrary macro works on enums with a single variant
+//! See https://github.com/model-checking/kani/issues/4170
+
+#[allow(unused)]
+#[derive(kani::BoundedArbitrary)]
+enum Foo {
+    A(#[bounded] String),
+}
+
+#[kani::proof]
+#[kani::unwind(6)]
+fn check_enum() {
+    let any_enum: Foo = kani::bounded_any::<_, 4>();
+    let Foo::A(s) = any_enum;
+    kani::cover!(s.len() == 0);
+    kani::cover!(s.len() == 1);
+    kani::cover!(s.len() == 2);
+    kani::cover!(s.len() == 3);
+    kani::cover!(s.len() == 4);
+}

